ibis/backends/clickhouse/compiler/core.py

```diff
@@ -0,0 +1,102 @@
+"""ClickHouse ibis expression to sqlglot compiler.
+
+The compiler is built with a few `singledispatch` functions:
+
+    1. `translate` for table expressions
+    1. `translate` for table nodes
+    1. `translate_rel`
+    1. `translate_val`
+
+## `translate`
+
+### Expression Implementation
+
+The table expression implementation of `translate` is a pass through to the
+node implementation.
+
+### Node Implementation
+
+There's a single `ops.Node` implementation for `ops.TableNode`s instances.
+
+This function:
+
+    1. Topologically sorts the expression graph.
+    1. Seeds the compilation cache with in-degree-zero table names.
+    1. Iterates though nodes with at least one in-degree and places the result
+       in the compilation cache. The cache is used to construct `ops.TableNode`
+       keyword arguments to the current translation rule.
+
+## `translate_rel`
+
+Translates a table operation given already-translated table inputs.
+
+If a table node needs to translate value expressions, for example, an
+`ops.Aggregation` that rule is responsible for calling `translate_val`.
+
+## `translate_val`
+
+Recurses top-down and translates the arguments of the value expression and uses
+those as input to construct the output.
+"""
+
+from typing import Any, Mapping
+
+import sqlglot as sg
+
+import ibis.expr.operations as ops
+from ibis.backends.clickhouse.compiler.relations import translate_rel
+from ibis.common.graph import Graph
+
+_SENTINEL = object()
+
+
+def translate(op: ops.TableNode, params: Mapping[ops.Node, Any]) -> sg.exp.Expression:
+    """Translate an ibis operation to a sqlglot expression.
+
+    Parameters
+    ----------
+    op
+        An ibis `TableNode`
+    params
+        A mapping of nodes to concrete values
+
+    Returns
+    -------
+    sqlglot.expressions.Expression
+        A sqlglot expression
+    """
+
+    dag = Graph.from_dfs(op, filter=ops.TableNode)
+
+    # get table nodes only, value expressions are translated top down
+    # within the various relations
+    #
+    # TODO(cpcloud): can values be translated bottom up as well?
+    nodes = list(dag.toposort())
+
+    # generate aliases for tables that don't have names already
+    # don't include the root ([:-1]) because it's not necessary to alias
+    aliases = {node: f"t{i:d}" for i, node in enumerate(nodes[:-1]) if dag[node]}
+
+    cache = {}
+
+    for node in nodes:
```

Comment:
I wonder, could we use `node.map()` here? Seemingly they serve similar purposes.

Reply:
This really reminds me of `Node.map()`'s implementation, could you please try to use that here?